sam.c

```diff
@@ -2383,8 +2383,155 @@ int sam_hdr_change_HD(sam_hdr_t *h, const char *key, const char *val)
  *** SAM record I/O ***
  **********************/
 
-static int sam_parse_B_vals(char type, uint32_t n, char *in, char **end,
-                            char *r, bam1_t *b)
+// The speed of this code can vary considerably depending on minor code
+// changes elsewhere as some of the tight loops are particularly prone to
+// speed changes when the instruction blocks are split over a 32-byte
+// boundary.  To protect against this, we explicitly specify an alignment
+// for this function.  If this is insufficient, we may also wish to
+// consider alignment of blocks within this function via
+// __attribute__((optimize("align-loops=5"))) (gcc) or clang equivalents.
+// However it's not very portable.
+// Instead we break into separate functions so we can explicitly specify
+// use __attribute__((aligned(32))) instead and force consistent loop
+// alignment.
+static inline int64_t grow_B_array(bam1_t *b, int64_t n, size_t size) {
```

Comment:
This interface could be simplified a bit by passing `n` as a pointer (and possibly also as a `size_t`), returning 0 on success.  The callers already take a pointer to the value, so wouldn't need to dereference it.

There's no check for overflow when calculating `bytes`.  It might be possible to get this to happen on 32-bit platforms by creating a tag of the form `bb:B:I:1000000,,,,,,` with a lot of commas.  It will expand to 4x the size of the textual version, which might be enough to do the trick before you run out of virtual address space.  The best place for such a test might be in `sam_parse_Bi_vals()` and `sam_parse_BI_vals()` where it can be specialised for a 4-byte `size`.

I don't think adding `sizeof(uint32_t)` to `bytes` is necessary here.  It was done in the original code to account for the length bytes, but they've already been factored in as this point.